pkg/visitors/createfv.go

```diff
@@ -0,0 +1,240 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package visitors
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"strconv"
+	"unsafe"
+
+	"github.com/linuxboot/fiano/pkg/guid"
+	"github.com/linuxboot/fiano/pkg/uefi"
+)
+
+// CreateFV creates a firmware volume at given offset
+type CreateFV struct {
+	AbsOffset uint64
+	Size      uint64
+	Name      guid.GUID
+
+	offset uint64
+	found  bool
+}
+
+// Run wraps Visit and performs some setup and teardown tasks.
+func (v *CreateFV) Run(f uefi.Firmware) error {
+	err := f.Apply(v)
+	if err != nil {
+		return err
+	}
+	if !v.found {
+		return fmt.Errorf("Cannot create FV at %#x (+%#x) no BIOS region found", v.AbsOffset, v.Size)
+	}
+	return nil
+}
+
+// Visit applies the Remove visitor to any Firmware type.
+func (v *CreateFV) Visit(f uefi.Firmware) error {
+	if v.found {
+		return nil
+	}
+	var offset, end uint64
+	switch f := f.(type) {
+	case *uefi.BIOSRegion:
+		if f.FRegion != nil {
+			offset = uint64(f.FRegion.BaseOffset())
+			end = uint64(f.FRegion.EndOffset())
+		} else {
+			end = f.Length
+		}
+		if v.AbsOffset < offset {
+			return fmt.Errorf("Cannot create FV at %#x, BIOS region starts at %#x", v.AbsOffset, offset)
+		}
+		if v.AbsOffset+v.Size >= end {
+			return fmt.Errorf("Cannot create FV ending at %#x (%#x + %#x), BIOS region ends at %#x", v.AbsOffset+v.Size, v.AbsOffset, v.Size, end)
+		}
+
+		// Manually visit children, we want the index
+		idx := -1
+		var bp *uefi.BIOSPadding
+	L:
+		for i, e := range f.Elements {
+			switch f := e.Value.(type) {
+			case *uefi.BIOSPadding:
+				bp = f
+				bpOffset := offset + f.Offset
+				bpEnd := bpOffset + uint64(len(f.Buf()))
+				if v.AbsOffset >= bpOffset && v.AbsOffset+v.Size <= bpEnd {
+
+					idx = i
+					break L
+				}
+			}
+		}
+
+		if idx == -1 {
+			return fmt.Errorf("Cannot create FV at %#x (+%#x) no matching BIOS Pad found", v.AbsOffset, v.Size)
+		}
+		v.found = true
+		fv, err := createEmptyFirmwareVolume(v.AbsOffset-offset, v.Size, &v.Name)
+		if err != nil {
+			return err
+		}
+		return insertFVinBP(f, v.AbsOffset-offset, bp, idx, fv)
+	}
+
+	return f.ApplyChildren(v)
+}
+
+func insertFVinBP(br *uefi.BIOSRegion, offset uint64, bp *uefi.BIOSPadding, idx int, fv *uefi.FirmwareVolume) error {
+	// Copy the Elements before the modified BIOS Pad
+	newElements := make([]*uefi.TypedFirmware, idx, len(br.Elements))
+	copy(newElements, br.Elements[:idx])
+
+	bpBuf := bp.Buf()
+	// keep part of the BIOS Pad before the new fv if needed
+	if bp.Offset < offset {
+		hbp, err := uefi.NewBIOSPadding(bpBuf[:offset-bp.Offset], bp.Offset)
+		if err != nil {
+			return err
+		}
+		newElements = append(newElements, uefi.MakeTyped(hbp))
+	}
+
+	// Add the FV
+	newElements = append(newElements, uefi.MakeTyped(fv))
+
+	// keep part of the BIOS Pad after the new fv if needed
+	if offset-bp.Offset+fv.Length < uint64(len(bpBuf)) {
```

Comment:
you can assign offset-bp.Offset+fvLength to a vairable and use it in the next line too.

Reply:
call it something like tbpStart or something

Reply:
done